jmsSource/src/main/java/io/vantiq/extsrc/jmsSource/JMSCore.java

```diff
@@ -0,0 +1,344 @@
+                return;
+            }
+            
+            // Getting the message format if it was specified, or defaulting to Message
+            if (request.get("JMSFormat") instanceof String) {
+                msgFormat = (String) request.get("JMSFormat");
+            } else {
+                log.debug("No JMSFormat was specified, the default Message message type will be used.");
+                msgFormat = "Message";
+            }
+            
+            // Sending the message to the appropriate destination
+            try {
+                localJMS.produceMessage(msg, dest, msgFormat, isQueue);
+            } catch (JMSException e) {
+                log.error("An error occured when attempting to send the given message.", e);
+            } catch (DestinationNotConfiguredException e) {
+                log.error("An error occured when attempting to send the given message. The source was not configured "
+                        + "to send messages to the following destination: " + dest + ".", e);
+            } catch(UnsupportedJMSMessageTypeException e) {
+                log.error("An error occured when attempting to send the given message. The provided JMS Message Type: "
+                        + e.getMessage() + " is either invalid or not currently supported.");
+            } catch (Exception e) {
+                log.error("An unexpected error occured when attempting to send the given message.", e);
+            }
+        }
+    }
+    
+    /**
+     * Called by the queryHandler. Used to read the next available message from the specified JMS Queue. The message is
+     * converted to JSON, and sent back to VANTIQ as a queryResponse.
+     * @param message   The Query message
+     */
+    public void readQueueMessage(ExtensionServiceMessage message) {
+      Map<String, ?> request = (Map<String, ?>) message.getObject();
+      String replyAddress = ExtensionServiceMessage.extractReplyAddress(message);
+      
+      // Get local copy of JMS
+      JMS localJMS;
+      synchronized (SYNCH_LOCK) {
+          localJMS = jms;
+      }
+      
+      // Make sure JMS is safe to use (has not been closed)
+      if (localJMS == null) {
+          if (client != null) {
+              client.sendQueryError(replyAddress, this.getClass().getName() + ".closed",
+                      "JMS connection closed before operation could complete.", null);
+          } else {
+              log.error("JMS connection closed before operation could complete.");
+          }
+      } else {
+          // Retrieve most recent message from specified queue. If no queue name is specified, or if exception is thrown, return query error.
+          if (request.get("queue") instanceof String) {
+              String queue = (String) request.get("queue");
+              try {
+                  Map<String, Object> messageMap = localJMS.consumeMessage(queue);
```

Comment:
Fred, Namir,
localJMS could be closed here, since close() could have been called after the SYNCH block above.
I guess one of the catch() clauses below will handle that case?

Reply:
This is indeed possible. We catch all of the exceptions that we expect to encounter, but also have a ```catch (Exception e)``` block at the end in case something unexpected happens. 

I think this could be considered an "unexpected" error, since it is pretty unlikely that the JMS instance would be closed between checking that it is non-null and actually using it to consume a queue message. Let me know if this is an inaccurate assumption.

Reply:
@namirfa I think it's expected in the sense that it's a documented exception that could come back.  Someone well could close it.  However, I think you're correct that the, to answer @rblumer4's question, we understand that that's quite possible, and the expectation is that we'll catch the exception.

The alternative, as you know, was to block access during the entirety of any JMS operation.  And since those may involve network or other/separate server access, that seems like a bad idea.

Reply:
that sound ok to me, too.